lib/ads.js

```diff
@@ -266,88 +288,72 @@ var readState = function (cb) {
-    readLength += oneOption.readLength
-    symNames.push({symname: oneOption.name, length: oneOption.readLength})
-  })
 
-  var options = {
-    indexGroup: ADSIGRP.SUMUP_READ,
-    indexOffset: commandOptions.length,
-    writeBuffer: buf,
-    readLength: readLength + commandOptions.length * 4,
-    symname: symNames
-  }
-
-  writeReadCommand.call(ads, options, function (error, result) {
-    if (error) {
-      cb(ads, error)
-    } else if (result.length > 0) {
-      var length = result.length
-      var curIndex = 0
-      var readResults = {}
-      _.each(commandOptions, function (oneOption, index) {
-        if (curIndex + 4 < length) {
-          readResults[oneOption.name] = {state: result.readUInt32LE(curIndex)}
-          curIndex += 4
-        }
+  getHandles.call(ads, commandOptions, function (err, handles) {
+    if (!err) {
+      var readLength = 0
+      var buf = Buffer.alloc(handles.length * 12)
+
+      handles.forEach(function (handle, index) {
+        buf.writeUInt32LE(ADSIGRP.RW_SYMVAL_BYHANDLE, index * 12 + 0)
+        buf.writeUInt32LE(handle.symhandle, index * 12 + 4)
+        buf.writeUInt32LE(handle.bytelength.length, index * 12 + 8)
+        readLength += handle.bytelength.length + 4
       })
-      _.each(commandOptions, function (oneOption, index) {
-        var buf
-        switch (oneOption.readLength) {
-          case 1:
-            if (curIndex + 1 <= length) {
-              buf = new Buffer(1)
-              result.copy(buf, 0, curIndex, curIndex + 1)
-              curIndex += 1
-            }
-            break
-          case 2:
-            if (curIndex + 2 <= length) {
-              buf = new Buffer(2)
-              result.copy(buf, 0, curIndex, curIndex + 2)
-              curIndex += 2
-            }
-            break
-          case 4:
-            if (curIndex + 4 <= length) {
-              buf = new Buffer(4)
-              result.copy(buf, 0, curIndex, curIndex + 4)
-              curIndex += 4
-            }
-            break
-          case 8:
-            if (curIndex + 8 <= length) {
-              buf = new Buffer(8)
-              result.copy(buf, 0, curIndex, curIndex + 8)
-              curIndex += 8
+    }
+
+    var request = {
+      indexGroup: ADSIGRP.SUMUP_READ,
+      indexOffset: commandOptions.length,
+      writeBuffer: buf,
+      readLength: readLength,
+      symname: 'multiRead',
+    }
+
+    writeReadCommand.call(ads, request, function (err, result) {
+      if (err) {
+        cb.call(ads, err)
+      } else {
+        if (result.length > 0) {
+          var resultpos = 0
+          var handlespos = commandOptions.length * 4
+
+          commandOptions.forEach(function (handle) {
+            if (!handle.err) {
+              var adsError = result.readUInt32LE(resultpos)
+              resultpos += 4
+
+              if (adsError != 0) {
+                handle.err = getError(adsError)
+              }
```

Comment:
if we export the "const errors = {" structure, we could just write the error code here. This would save memory for many errors. but we should then describe in readme.md how to evaluate it.

Reply:
handle.err = getError(adsError)